### PR TITLE
Add Search Capability and Introduction of Search Tags to workshops

### PR DESCRIPTION
**Pull request by bajpai244**
The following pull request adds search capability to the workshops along with **another feature Search Tags.**

test it live at: [https://workshops-1xheizjaa.hackclub.dev/](https://workshops-1xheizjaa.hackclub.dev/) 

### What are these Tags? 

Tags are YAML attributes that you add to your workshop's .md files to help people discover them more easily via the search option being integrated into the workshop.

For example, I have a hardware workshop, then I can use the following tags to make it more searchable.

(in your .md's YAML)
`tags:  Arduino, IoT, System on a chip, AI, Machine Learning, Deep Learning`

Yes! **Tags are comma-separated.**

Now, to prevent people from tricking the system, **only the first 5 tags** are considered by the search layer to search the workshop.

_Meaning from the above tags Deep Learning will not be considered as it is the 6th tag._

### So how will the search layer work?

The search layer is basically a bridge that filters data on the basis of the search queries.

It takes into consideration the following data attributes to make a workshop eligible on any particular search query:

1. Section Title
2. Section Description
3. Workshop Title
4. Workshop Description
5. **Workshop' Search Tags**

So if any of the following attributes have a match for the search query then the associated workshop or section becomes eligible for the result.

_All the commits have a proper explanation of the changes in the codebase and why they were made._

The following code guarantees support for:
1. All Major Browsers
2. Different Screen Sizes ( responsive )
3. Complete Search working on the frontend ( meaning it is super fast 🚄 )
4. Backward Compatibilty ( no breaking changes )

This is my small effort to make workshops more searchable and discoverable :)

### CODE REVIEW

**Comment by lachlanjc** [comment]
#### 📝Changed routes:


| [**/\**](https://workshops-hihkvfvyb.hackclub.dev/\) | [**/**](https://workshops-hihkvfvyb.hackclub.dev/) |
|:-:|:-:|
|<a href="https://workshops-hihkvfvyb.hackclub.dev/\">Screenshot of /\</a><br /><sup><a href="https://images.weserv.nl?url=https://iad.microlink.io/fG8wdvtXwCSZIC7H98690pA0d92tauUgLaiUMWfFOpaT3OvawB4thaphvnpI2DTqPNElwOOesz5lDffaYC5Zcg.png">(view full size)</a> |<a href="https://workshops-hihkvfvyb.hackclub.dev/">Screenshot of /</a><br /><sup><a href="https://images.weserv.nl?url=https://iad.microlink.io/MSVwJMkjnpRDN0QD9NK023j3yhbyNPJhdLsbgQFgcjJIhXEy8TZnv4QiFVqmQrBcAIH91d0n63Xg8dWeFEh8AA.png">(view full size)</a> |







Commit 91b252abaccc8b7e5674d7c53ffda14470e6a883 (https://workshops-hihkvfvyb.hackclub.dev).

File: components/nav.js
Comment (sampoder) [inline components/nav.js:217]: Could this input be more inline with the Hack Club theme?

Specifically having less border radius and changing based on light/dark mode. Happy to do this if needed.

File: components/nav.js
Comment (sampoder) [inline components/nav.js:191]: might want to remove this

**Review by sampoder — CHANGES_REQUESTED** [review]
This is so cool! Hope it's alright if I give some feedback.

One thing I noticed is when I'm playing the colourful clicking thing on the bg of the header and then search all my colours go away

File: components/nav.js
Comment (bajpai244) [reply components/nav.js:217]: Hey Sam 👋  

So, I would highly encourage you in helping with the dark theme colors for the input tag, Do you have any suggestion on what color it can be?

For the border-radius:

so if we reduce the border radius then this is what it will look like:

image

This is why I preferred it bo more rounded

But if you think that having it less rounded makes it look better then do lemme know :)

File: components/nav.js
Comment (bajpai244) [reply components/nav.js:191]: Hey, yeah I was using this statement for some debugging, lemme remove that :)

Thanks, you are awesome 🙌

**Comment by bajpai244** [comment]
Hey @sampoder , 

I highly encourage feedback on this :)

can you provide me a screenshot so that I can better understand it?

File: components/nav.js
Comment (sampoder) [reply components/nav.js:217]: If you're using CSS I think `var(--theme-ui-colors-background)` would work best

I personally prefer the less rounded variant though I acknowledge these things are highly subjective, @maxwofford @MatthewStanciu thoughts?

File: components/nav.js
Comment (sampoder) [reply components/nav.js:191]: Anytime

**Comment by sampoder** [comment]
Here's a video of the colour bug:

https://user-images.githubusercontent.com/39828164/118930219-0bcb0280-b978-11eb-8d95-f40513ff71e4.mov

**Comment by bajpai244** [comment]
@sampoder got it,

It is re-rendering when the state changes, I am onto fixing it

Would highly appreciated if u too can look into what is actually trigerring it's re-render so that we can fix it ASAP 🚀

**Comment by bajpai244** [comment]
@sampoder this commit fixes everything we have discussed till this.

1. Now, the Blog doesn't re-render on searching :)
2. The input tag is theme-aware

**Comment by bajpai244** [comment]
@maxwofford @MatthewStanciu  what do u people think about the border-radius of input?

@sampoder **prefers it to not be circular,** if u people think that what sam is proposing will look better then do let me know I will make the desired changes 🙌

**Comment by bajpai244** [comment]
The new deployment link: [https://workshops-1xheizjaa.hackclub.dev/](https://workshops-1xheizjaa.hackclub.dev)

File: package.json
Comment (maxwofford) [inline package.json:33]: What are we using this for? I don't see any usage of sharp in the code and nothing else in yarn.lock depends on it.

If you do need it, heads up that it's currently not working in ARM Macs, so let's use an alternative if we can.

File: pages/\
Comment (maxwofford) [inline pages/\:1]: This file looks funny-- is it intentional?

File: lib/search.js
Comment (maxwofford) [inline lib/search.js:35]: I'll defer to @MatthewStanciu on code formatting, but want to ping him here b/c this follows different formatting then the rest of the codebase.

Comment (maxwofford) [reply lib/search.js:35]: @bajpai244 feel free to resolve this if he doesn't comment here.

**Review by maxwofford — CHANGES_REQUESTED** [review]
This is so awesome. Thanks for the contribution @bajpai244!

I've left a couple comments on different lines. Additionally, there are a couple .DS_Store files committed on accident. Could you please remove them? [This is what I personally use](https://sebastiandedeyne.com/setting-up-a-global-gitignore-file/).

**Comment by maxwofford** [comment]
Screen Shot 2021-05-20 at 10 12 07

When I click this, I expect it to clear my search query & show me all the workshops. Currently it just hides the input, but still filters by my search term.

File: pages/\
Comment (bajpai244) [reply pages/\:1]: It is required in the new version of theme-ui :)

It helps u use the new automatic runtime being provided by  React

File: package.json
Comment (bajpai244) [reply package.json:33]: **Hey, it does work in arm macs, I have tested it on mine :)**

So when I ran the project, it required me to install sharp, it was saying that can't find sharp, so have to import it to make it work.

I think some package is using it as a dependency, but still, it does work on arm Macs.

File: pages/\
Comment (bajpai244) [reply pages/\:1]: so when we use it, it uses theme-ui to transform the JSX ( which is required to use sx prop in custom components )

File: lib/search.js
Comment (bajpai244) [reply lib/search.js:35]: I see, lemme run the prettier.config file through my whole codebase to resolve this

**Comment by bajpai244** [comment]
Hey @maxwofford, thanks for reporting this, lemme me fix that ASAP.

File: pages/\
Comment (sampoder) [reply pages/\:1]: yep, you're right but I don't think that's the problem, check the file name

Comment (bajpai244) [reply pages/\:1]: Next js has sharp as an optional dependency, but sometimes not having it breaks it ( what happened with me ) :

https://github.com/vercel/next.js/issues/20456#:~:text=Describe%20the%20bug-,Next.,without%20sharp%20present%20in%20node_modules.

File: package.json
Comment (sampoder) [reply package.json:33]: That's weird

**Comment by bajpai244** [comment]
Hey @maxwofford, have fixed everything u reported and I have also formatted all my files using the project's prettier.config.js 🙌

File: package.json
Comment (bajpai244) [reply package.json:33]: found out, next js has an optional dependency on it, but sometimes not having it causes it to break ( what happened in my case )

Comment (sampoder) [reply package.json:33]: interesting, probably related to the image optimization

**Comment by bajpai244** [comment]
hey @maxwofford, I have also removed .DS_STORE file that was committed by mistake, so have till this far, I have resolved everything reported, If I am missing on something then do let me know

File: package.json
Comment (bajpai244) [reply package.json:33]: yeah :)

File: pages/\
Comment (bajpai244) [reply pages/\]: deleted, lol! don't know how it got created

**Comment by bajpai244** [comment]
> Screen Shot 2021-05-20 at 10 12 07
> 
> When I click this, I expect it to clear my search query & show me all the workshops. Currently, it just hides the input, but still filters by my search term.

done 🚀

File: package.json
Comment (maxwofford) [reply package.json:33]: It works on my m1 by failing to install sharp & continuing without it as a dependency:

Screen Shot 2021-05-20 at 15 50 19

Comment (bajpai244) [reply package.json:33]: So, when I remove sharp this is what I get:
image

And this is how my page looks
image

Now very strangely, when I install sharp, although I just saw that its build failed ( **sorry for saying earlier that it build successfully, I just noticed it** ), but installing it **_somehow fixes everything for me_**

Comment (bajpai244) [reply package.json]: Hey @maxwofford, just saw your PR and It has fixed this issue, I have merged it and now we longer have sharp as a dependency 🚀 

Thanks you are awesome 🙌

**Comment by bajpai244** [comment]
@MatthewStanciu, we have also removed **sharp** as a dependency, thanks to @maxwofford for his PR 🙌

**Comment by bajpai244** [comment]
@MatthewStanciu, so there is a small conflict on the yarn.lock file after I merged the PR from @maxwofford, can any of you resolve it 👀

**Comment by maxwofford** [comment]
Double check `public/.DS_Store`

**Comment by bajpai244** [comment]
@maxwofford I checked it and I have removed it, see the ss below

image

**Comment by MatthewStanciu** [comment]
Thanks so much for your work Harsh 💖 I'm so glad we can finally search the workshops page 🎉

**Comment by bajpai244** [comment]
Thanks, @MatthewStanciu, @maxwofford, and @sampoder for being so helpful.

You people are amazing :)

**Pull request merged**
